packages/@lwc/template-compiler/src/shared-next/types.ts

```diff
@@ -0,0 +1,187 @@
+    True = 'true',
+    False = 'false',
+}
+
+export interface BaseNode {
+    type: string;
+    location: SourceLocation;
+}
+
+export interface SourceLocation {
+    startLine: number;
+    startColumn: number;
+    endLine: number;
+    endColumn: number;
+    start: number;
+    end: number;
+}
+
+export interface Literal<Value = string | boolean> {
+    type: LWCNodeType.Literal;
+    value: Value;
+}
+
+export interface Identifier extends BaseNode {
+    type: LWCNodeType.Identifier;
+    name: string;
+}
+
+export interface MemberExpression extends BaseNode {
+    type: LWCNodeType.MemberExpression;
+    object: Expression;
+    property: Identifier;
+}
+
+export type Expression = Identifier | MemberExpression;
+
+export interface Attribute extends BaseNode {
+    type: LWCNodeType.Attribute;
+    name: string;
+    value: Literal | Expression;
+}
+
+export interface Property extends BaseNode {
+    type: LWCNodeType.Property;
+    name: string;
+    value: Literal | Expression;
+}
+
+export interface EventListener extends BaseNode {
+    type: LWCNodeType.EventListener;
+    name: string;
+    handler: Expression;
+}
+
+export interface Directive extends BaseNode {
+    type: LWCNodeType.Directive;
+    name: string;
+    value: Expression | Literal;
+}
+
+export interface KeyDirective extends Directive {
+    name: LWCNodeType.Key;
+    value: Expression;
+}
+
+export interface DynamicDirective extends Directive {
+    name: LWCNodeType.Dynamic;
+    value: Expression;
+}
+
+export interface RenderModeDirective extends Directive {
+    name: LWCNodeType.RenderMode;
+    value: Literal<LWCDirectiveRenderMode.Shadow> | Literal<LWCDirectiveRenderMode.Light>;
+}
+
+export interface PreserveCommentsDirective extends Directive {
+    name: LWCNodeType.PreserveComments;
+    value: Literal<boolean>;
+}
+
+export type ElementDirective = KeyDirective | DynamicDirective;
+export type RootDirective = RenderModeDirective | PreserveCommentsDirective;
+
+export interface Text extends BaseNode {
+    type: LWCNodeType.Text;
+    value: Literal | Expression;
+}
+
+export interface Comment extends BaseNode {
+    type: LWCNodeType.Comment;
+    value: string;
+}
+
+export interface BaseParentNode extends BaseNode {
+    children: ChildNode[];
+}
+
+export interface Element extends BaseParentNode {
+    type: LWCNodeType.Element;
+    name: string;
+    namespace?: string;
+    attributes: Attribute[];
+    listeners: EventListener[];
+    directives?: ElementDirective[];
+}
+
+export interface Component extends BaseParentNode {
+    type: LWCNodeType.Component;
+    name: string;
+    attributes: Attribute[];
+    properties: Property[];
+    listeners: EventListener[];
+    directives?: ElementDirective[];
+}
+
+export interface IfBlock extends BaseParentNode {
+    type: LWCNodeType.IfBlock;
+    modifier: LWCIfBlockModifier.True | LWCIfBlockModifier.False;
+    condition: Expression;
+}
+
+export interface ForBlock extends BaseParentNode {
```

Comment:
The new AST shape only has a `ForBlock` which satisfies the requirements for the previous [ForEach](https://github.com/salesforce/lwc/blob/66a26f6a8863905d60630bf604cc4761ac345c60/packages/%40lwc/template-compiler/src/shared/types.ts#L32) type but how should we handle when the for loop is an iterator?

Previously we had the [ForIterator](https://github.com/salesforce/lwc/blob/66a26f6a8863905d60630bf604cc4761ac345c60/packages/%40lwc/template-compiler/src/shared/types.ts#L38) type to handle iterators but it's not in the new shape. I can set the iterator name to the item property on the `ForBlock` type but I don't think there's an easy way to distinguish between iterators and for each this way.

Alternatively I can create `ForEach` and `ForIterator` subtypes from `ForBlock` but wanted to get your thoughts on this.

Distinguishing between the ForEach and ForIterator comes up during [codegen](https://github.com/salesforce/lwc/blob/66a26f6a8863905d60630bf604cc4761ac345c60/packages/%40lwc/template-compiler/src/codegen/index.ts#L101) and also when [traversing](https://github.com/salesforce/lwc/blob/66a26f6a8863905d60630bf604cc4761ac345c60/packages/%40lwc/template-compiler/src/parser/index.ts#L904) through the parent stack.

Reply:
> Alternatively I can create `ForEach` and `ForIterator` subtypes from `ForBlock` but wanted to get your thoughts on this.

Sounds good to me. Feel free to make adjustments to the AST shape as needed. We can review the changes prior merging the PR.